### ISSUE TITLE
RF01 false positive reported for fields in group by, having clauses

### ISSUE THREAD

**Issue by AnithaG-Oak** [sqlfluff/sqlfluff#6527]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Sqlfluff reports `RF01: references.from` error for valid queries when there are joins and group by/having clause are used.

### Expected Behaviour

No sqlfluff issues reported for the valid queries.

### Observed Behaviour

Error reported:
`Reference 'total' refers to table/view not found in the FROM clause or found in ancestor statement.sqlfluff(RF01: references.from)`

### How to reproduce

`with left_table as (
    select
        'a' as id,
        10 as value
),

right_table as (
    select
        'a' as id,
        20 as value
),

final as (
    select
        left_table.id,
        (left_table.value + right_table.value) as total
    from left_table
    left join right_table
        on left_table.id = right_table.id
    having total > 10  -- Same error reported for group by as well 
)

select * from final
`

### Dialect

BigQuery

### Version

3.3.0

### Configuration

[sqlfluff]
dialect = bigquery
templater = dbt
exclude_rules = L029, L034, L044, ST03
max_line_length = 120

[sqlfluff:rules:layout.long_lines]
ignore_comment_lines = False
ignore_comment_clauses = False

[sqlfluff:rules:capitalisation.keywords]
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.identifiers]
extended_capitalisation_policy = lower

[sqlfluff:rules:capitalisation.functions]
extended_capitalisation_policy = lower

[sqlfluff:rules:capitalisation.literals]
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.types]
extended_capitalisation_policy = lower

[sqlfluff:templater:jinja]
apply_dbt_builtins = true

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by Dzhmilka** [comment]
The issue here is not with the linter, but with how some aliases are handled

Simply saying, the program doesn't see that alias yet as you just defined it
Try moving this logic to the select statement, like this:

```
select * from final
having total > 10
```

**Comment by rbeilvert** [comment]
@Dzhmilka the alias exists for a `having` or `group by`. The example you show is about using an alias in a `where` clause so it's not relevant.
Your link shows the use of chatgpt... Please be careful with that 😶

**Comment by Dzhmilka** [comment]
Hello, [@rbeilvert]
The link was indeed obtained from ChatGPT, and you are right - the topic, that is being discussed there, differs slightly from what has been sent here.

But my point was, that the query sent there is syntactically incorrect. If you try to run it in BigQuery environment - it will fail on its own.
Maybe that person accidentally sent slightly incorrect code, and the bug actually exists even with the correct version, but I wanted to point this out, because wasn't sure.

I've corrected this a little bit:
```
with left_table as (
    select
        'a' as id,
        10 as value
),

right_table as (
    select
        'a' as id,
        20 as value
),

final as (
    select
        left_table.id,
        sum(left_table.value + right_table.value) as total
    from left_table
    left join right_table
        on left_table.id = right_table.id
    group by left_table.id
    having total > 10
)

select * from final
```

**Comment by PaddyAlton** [comment]
FWIW I think this is a real bug (I imagine @rbeilvert was creating a minimal reproducible example and left out the `group by` by accident).

I came across a very similar example in my work: two tables joined together, aggregate function over a column from one of them, `RF01` failure for `HAVING aggregated_field > some_value`.

The query runs just fine and it's not possible to give `aggregated_field` a reference since it is created in the same query.

**Comment by annebelleo** [comment]
I believe this is fixed by #7188. I tested all queries listed here locally with `sqlfluff=3.5.0` and don't get an RF01 error.
